### PR TITLE
[Issue 199] Add extra hidden cell add buttons

### LINKED ISSUE #199: More options for plus button when adding cell to layout
New options. Could be implemented in a drop down from the + icon on each cell.

* Add to top of layout (full width bottom row, default)
* Add to top of dashboard layout (full width top row)
* Add to relative notebook position

Last one is tricky because what if the user has put a cell that appears lower in the notebook flow higher in the dashboard flow than other cells? Maybe the algorithm can be:

1. Degenerate case, if the dashboard layout is empty, just add the cell full width.
2. Get the notebook flow index of the cell being added to the flow. Call this `i`.
3. Search top-left to bottom-right in reading order for the cell with the closest notebook index `j` that does not exceed `i`.
4. Insert the `i`th cell in a new full width row directly below the row containing the `j`th cell. (Why full width? Avoids a tiny cell shove into an existing row with content overflowing and obscuring the controls to make further adjustments.)

### PR DESCRIPTION
**Pull request by dalogsdon**
Add to top
Add to bottom
Add relative to notebook position

Resolves #199

### CODE REVIEW

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js [hunk 1]
```diff
@@ -400,10 +422,28 @@ define([
...
+                });
+            gc.find('.add-btn-relative').click(function() {
+                    // show a cell full width relative to notebook position
+                    var $cell = self._getParentCell(this);
+                    var i = $cell.index();
+                    var $prevCell = $cell.parent()
+                        .find('.cell')
+                        .slice(0,i)
+                        .filter('.grid-stack-item')
+                        .last();
```
Comment (jhpedemonte) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js]: `$cell.prev(SELECTOR)` doesn't work ([docs](http://api.jquery.com/prev/))?

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js [hunk 2]
```diff
@@ -605,7 +647,9 @@ define([
         this.$container.one('transitionend',
             containerTransitionEnd.resolve.bind(containerTransitionEnd));
         $.when(cellTransitionEnd, containerTransitionEnd)
-            .then(this._repositionHiddenCells.bind(this));
+            .then(function() {
+                this._repositionHiddenCells();
+            }.bind(this));
```
Comment (jhpedemonte) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js]: Out of curiosity, why this change?

Comment (dalogsdon) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js]: That might be a leftover from refactoring. I think I had some other code in there at one point and then moved it out. I can fix it back to the original.

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js (hunk 1 above)
Comment (dalogsdon) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js]: I didn't realize `prev()` existed; it looks like `prevAll('.grid-stack-item').first()` is what I could use here as I need the closest previous non-hidden sibling.

**Comment by dalogsdon** [comment]
Updated per @jhpedemonte's comments.

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js [hunk 3]
```diff
@@ -390,7 +389,30 @@ define([
         '</div>' +
         '<div class="grid-control-container grid-control-ne">' +
             '<i class="grid-control close-btn fa fa-close fa-fw"></i>' +
-            '<i class="grid-control add-btn fa fa-plus fa-fw"></i>' +
+            '<span class="grid-control add-btn add-btn-relative fa-fw" title="Add to the dashboard according to its position in the notebook">' +
```
Comment (jhpedemonte) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js]: I think we can make the tooltip text a bit more concise: "Add to dashboard according to cell position in notebook"

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js [hunk 4]
```diff
@@ -390,7 +389,30 @@ define([
...
+                '<span class="bar-container">' +
+                    '<span class="bar bar-lighter"></span>' +
+                    '<span class="bar"></span>' +
+                    '<span class="bar bar-lighter"></span>' +
+                '</span>' +
+                '<i class="fa fa-plus"></i>' +
+            '</span>' +
+            '<span class="grid-control add-btn add-btn-top fa-fw" title="Add to the top of the dashboard">' +
```
Comment (jhpedemonte) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js]: "Add to top of dashboard"

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js [hunk 5]
```diff
@@ -390,7 +389,30 @@ define([
...
+                '<span class="bar-container">' +
+                    '<span class="bar"></span>' +
+                    '<span class="bar bar-lighter"></span>' +
+                    '<span class="bar bar-lighter"></span>' +
+                '</span>' +
+                '<i class="fa fa-plus"></i>' +
+            '</span>' +
+            '<span class="grid-control add-btn add-btn-bottom fa-fw" title="Add to the bottom of the dashboard">' +
```
Comment (jhpedemonte) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js]: "Add to bottom of dashboard"

File: jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js (hunk 3 above)
Comment (parente) [inline jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js]: "Add to dashboard in notebook cell order" ?

**Comment by jhpedemonte** [comment]
Tested locally, looks good to me. Only (minor) issue I had was wording of tooltips (see comments above).

**Comment by dalogsdon** [comment]
Updated titles.

**Comment by parente** [comment]
Trying it on the big UX survey dashboard (https://github.com/ibm-et/design/blob/analysis/surveys/2015-notebook-ux/analysis/report_dashboard.ipynb), the new options work as spec'ed. Nice.

In using them, a few more questions come to mind:

1. When I add a new cell that has output to a big notebook, it currently gets appended to the bottom of the layout. Should it instead appear in cell order as if I had clicked the middle + button option? I think this is closer to where I usually want the cell to appear than the end of the dashboard is.
2 If I want to quickly move a cell that is already in the layout to the top, bottom, or natural cell order, I first have to remove it from the layout in order to use the + buttons. That seems wrong. Should we have the same move shortcut buttons appear on cells already in the layout?
3. Right now, if I click to add a cell to its notebook position, I have to scroll around to find it and finish sizing it. In this case, I want the view to scroll automatically to the cell location in the layout. That said, sometimes I'm down in the hidden cells adding a bunch of them by click, click, clicking repeatedly. In this case, I do not want the view to scroll. If we do #2, I think I'd want the view to scroll to track the cell I clicked + on to move within the layout. I don't have a great idea on how to handle all these cases to perfection that doesn't also violate the rule of least surprise. Just brain dumping.

**Comment by dalogsdon** [comment]
1. That makes sense
2. Also makes sense
3. The scrolling is indeed tricky. I experimented a bit with the scrolling when adding hidden cells to the dashboard. There is one technical issue when animating the `scrollTop` property of the `#site` element at the same time as a cell is moving. I'm not sure if/how the scroll animation is getting interrupted. If we want to scroll on 'move to top/bottom' of cells as in 2 above, we may need to dig into this issue.

**Comment by jhpedemonte** [comment]
@parente: Agree with some of your suggestions. Thinking those should be handled in follow-up PR.

> Should it instead appear in cell order as if I had clicked the middle + button option?

Yes, I like that. Shouldn't be hard to handle.

> Should we have the same move shortcut buttons appear on cells already in the layout?

Sure, makes sense and would come in handy. We would also have to disable the relevant icon if the cell is already in that position (i.e. if at top, disable "move to top" button; might be interesting for cell order button).

> I want the view to scroll automatically to the cell location in the layout

@dalogsdon and I talked about this last week. My suggestion was to use a keyboard modifier on the buttons:
* **Default behavior:** Viewport stays at current position when clicking add. This allows adding multiple.
* **Shift + click:** Viewport scrolls to position of added cell.

**Comment by dalogsdon** [comment]
I will open 3 issues as discussed above.

**Comment by dalogsdon** [comment]
I opened #207, #208, #209 to handle the items brought up by @parente.

**Comment by parente** [comment]
If you guys want to merge this PR, go ahead. I'm fine with the initial cut and the others stuff in follow-ons.
